### PR TITLE
PEPPER-856 . Handle results file name for on-demand results activity creation

### PR DESCRIPTION
**Pull request by ssettipalli**
## Context

In the followup/OnDemand Activity Creation: Update the route to accept new parameter "resultsFileName" from DSM and use the value to populate an answer post activity instance creation for the DSS UI to use in results file download.

Opted for approach#2 in the ticket to minimize the changes to DSS core / platform.
Need to add special study / activity specific logic 

## FUD Score

_Overall, how are you feeling about these changes?_

- [x ] :relaxed: All good, business as usual!
- [ ] :sweat_smile: There might be some issues here
- [ ] :scream: I'm sweaty and nervous

## How do we demo these changes?

_How does one observe these changes in a deployed system? Note that **user visible** encompasses many personas--not just patients and study staff, but also ops duty, your fellow devs, compliance, etc._

- [x ] They are user-visible in dev as a regular user journey and require no additional instructions.
- [ ] Getting dev into a state where this is user-visible requires some tech fiddling. I have documented these steps in the related ticket.
- [ ] Requires other features before it's human visible. I have documented the blocking issues in jira.
- [ ] I have no idea how to demo this. Please help me!

## Testing

- [x] I have written automated positive tests
- [x] I have written automated negative tests
- [ ] I have written zero automated tests but have poked around locally to verify proper functionality
- [x ] The jira ticket has acceptance criteria and QA has the needed information to test changes

## Release

- [x ] These changes require no special release procedures--just code!
- [ ] Releasing these changes requires special handling and I have documented the special procedures in the release plan document

### CODE REVIEW

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/dsm/TriggerActivityPayload.java [hunk 1]
```diff
@@ -15,16 +15,29 @@ public class TriggerActivityPayload {
     @SerializedName("triggerId")
     private Long triggerId;
 
+    @SerializedName("resultsFileName")
```
Comment (ssettipalli) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/json/dsm/TriggerActivityPayload.java:18]: optional new param that applies to only lms/osteo and SOMATIC_RESULTS activity

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java [hunk 2]
```diff
@@ -101,6 +112,25 @@ public Object handle(Request request, Response response, TriggerActivityPayload
                 handle.attach(DataExportDao.class).queueDataSync(user.getId(), studyId);
                 log.info("Created on-demand activity instance {} for study guid {}, activity code {}, participant guid {}",
                         instanceDto.getGuid(), studyGuid, activityCode, participantGuid);
+
+                //create answer for this instance if lms/osteo2:SOMATIC_RESULTS
+                if ((studyGuid.equalsIgnoreCase("cmi-lms") || studyGuid.equalsIgnoreCase("CMI-OSTEO"))
```
Comment (ssettipalli) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java]: study and activity specific logic which is a negative but that minimizes core DSS platform changes including NO new schema changes.

Comment (otchet-broad) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java]: Why do we need the "study" portion of the conditional?  If we decide to add the RESULT_FILE_ACTIVITY_ID to other studies, this code should work, right?

Comment (ssettipalli) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java]: I debated about it and tried to be extra cautious. I will remove study check. None of existing studies have this activity any way

Comment (ssettipalli) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java]: removed study specific check and added unit tests

File: pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java [hunk 3]
```diff
@@ -228,6 +257,48 @@ public void test_reachedMaximumInstances_disallowTriggering() {
                 .body("message", containsString("maximum"));
     }
 
+    @Test
+    public void test_invalidFileName_failTriggering() {
```
Comment (ssettipalli) [inline pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java:262]: negative test case for new code

File: pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java [hunk 4]
```diff
@@ -228,6 +257,48 @@ public void test_reachedMaximumInstances_disallowTriggering() {
...
+                .body(new TriggerActivityPayload(userGuid, 9876L), ObjectMapperType.GSON) //no results file name passed
+                .when().post(url)
+                .then().assertThat()
+                .statusCode(422).contentType(ContentType.JSON)
+                .body("code", equalTo(ErrorCodes.ANSWER_NOT_FOUND))
+                .body("message", containsString("Invalid results file"));
+    }
+
+    @Test
+    public void test_successfulSomaticResultsTrigger() {
```
Comment (ssettipalli) [inline pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java:274]: positive test case for new code

**Comment by ssettipalli** [comment]
@denniscunningham 
just fyi: My observation, nothing critical or show stopper.
I have a feeling SONARCLOUD test coverage is not accurate and failing to recognize some scenarios where coverage exists.
For ex: I added 2 unit test which should my new changes but SONAR fails to recognize it.
May be the IF conditions with the value in the passed strings is not recognized by SONAR unless refactored to have new code in public / protected methods and called directly by unit tests.
cc @otchet-broad

File: pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java [hunk 5]
```diff
@@ -228,6 +258,48 @@ public void test_reachedMaximumInstances_disallowTriggering() {
...
+        given().auth().oauth2(dsmClientAccessToken)
+                .pathParam("activityCode", somaticResultsActivity.getActivityCode())
+                .body(new TriggerActivityPayload(userGuid, 9876L, "testSomaticResultsFile.pdf"), ObjectMapperType.GSON)
```
Comment (otchet-broad) [inline pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java:277]: Where does the magic 9876L come from?

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java [hunk 6]
```diff
@@ -101,9 +112,32 @@ public Object handle(Request request, Response response, TriggerActivityPayload
...
         return null;
     }
+
+    private void populateResultsFileNameAnswer(Response response, TriggerActivityPayload payload, String studyGuid,
+                                               String activityCode, Handle handle, User user, ActivityInstanceDto instanceDto) {
+        log.info("Populating answer for {} {}", studyGuid, activityCode);
+        String resultsFileName = payload.getResultsFileName();
+        if (StringUtils.isBlank(resultsFileName)) {
+            ApiError err = new ApiError(ErrorCodes.ANSWER_NOT_FOUND, "Invalid results file");
+            ResponseUtil.haltError(response, HttpStatus.SC_UNPROCESSABLE_ENTITY, err);
```
Comment (otchet-broad) [inline pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java]: Why wouldn't a 404 be appropriate here?

File: pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java (hunk 5 above)
Comment (ssettipalli) [reply pepper-apis/dss-core/src/test/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRouteTest.java:277]: DSM

File: pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java (hunk 6 above)
Comment (ssettipalli) [reply pepper-apis/dss-core/src/main/java/org/broadinstitute/ddp/route/DsmTriggerOnDemandActivityRoute.java]: updated

**Review by otchet-broad — APPROVED** [review]

**Pull request merged**